Support memory report logs with interleaved non-report output

vugarmammadli opened the pull request:
<!-- Please describe your changes on the following line: -->
`line 30-31` in `servo/etc/memory_reports_over_time.py`  should just ignore any line that doesn't begin with `|`.

The new if statement added such as `if line.startswith('|'):` then continue to append. So, only lines begin with `|` will be appended.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #20257(github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because of __

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20259)
<!-- Reviewable:end -->

vugarmammadli committed:
Support memory report logs with interleaved non-report output, added automated tests

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @jdm (or someone else) soon.

jdm commented:
Could you update the automated test in the file (run with `python ./etc/memory_reports_over_time.py --test`) to show this works?

vugarmammadli commented:
hey @jdm, 
Thanks for a comment. How should I update automated tests? Just running that line of code?

jdm commented:
@vugarmammadli The automated test exists in the same file. Look for the test function, and it has a string that is the input.

vugarmammadli commented:
@jdm thank you for your help! According to bug, it says it should ignore lines don't start with `|`. For test function, when I want to check the line without `|`, `parse_memory_report` function gives assertion error, because of the line `assert(line[0] == '|')`.

jdm commented:
@vugarmammadli Good point! We only test `parse_memory_report` right now; could you add a test for `extract_memory_reports` that contains a single report with a line that is not part of the report?

vugarmammadli commented:
@jdm Could you please provide any single report template for me? So I can use that as an input and add a line that is not part of the report for my test case.

jdm commented:
@vugarmammadli Run `./mach run -m 10` and copy the output from the terminal.

vugarmammadli commented:
hey @jdm,
thanks for help!

When I use the test case:

Begin memory reports
|
|  154.56 MiB -- explicit
|     107.88 MiB -- system-heap-unclassified
End memory reports

the output of extract_memory_reports() returns
`([['|', '|  154.56 MiB -- explicit', '|     107.88 MiB -- system-heap-unclassified']], ['reports'])`

when I use the same input without '|' in one line:

Begin memory reports
|
  154.56 MiB -- explicit
|     107.88 MiB -- system-heap-unclassified
End memory reports

it returns back
`([['|', '|     107.88 MiB -- system-heap-unclassified']], ['reports'])`

It seems like it ignores all lines without "|", but I don't think if the desired output should be like this.

jdm commented:
Why not? Given the input, that output looks correct to me.

vugarmammadli commented:
@jdm oh, I think I misunderstood it. Thanks a lot! Is that enough for test case?

In addition to this, I don't know why Taskcluster (pull_request) is failed..

jdm commented:
You can ignore Taskcluster; it always fails. That should be enough for the test case as long as it's automated!

vugarmammadli commented:
@jdm thank you!

For testing, I added `test_extract_memory_reports()` function, which tests `extract_memory_reports()` function.
In addition to this, I added `--test-extract-memory-reports` flag for running it from command line and added explanation to `usage()` function.

Please let me know if everything is fine!

jdm reviewed (changes requested):
sorry about the delay; this review was pending and I accidentally didn't post it.

jdm commented:
Additionally, be sure that `./mach test-tidy` passes with your changes.

jdm commented:
TravisCI reports the following `./mach test-tidy` failures:
```
./etc/memory_reports_over_time.py:75: E302 expected 2 blank lines, found 1
./etc/memory_reports_over_time.py:77: E128 continuation line under-indented for visual indent
./etc/memory_reports_over_time.py:78: E128 continuation line under-indented for visual indent
./etc/memory_reports_over_time.py:79: E128 continuation line under-indented for visual indent
./etc/memory_reports_over_time.py:80: E128 continuation line under-indented for visual indent
./etc/memory_reports_over_time.py:85: E302 expected 2 blank lines, found 1
```

vugarmammadli commented:
@jdm I guess it should not fail anymore after the last commit.

jdm commented:
Final request - could you [squash](https://www.devroom.io/2011/07/05/git-squash-your-latests-commits-into-one/) the four commits into one?

vugarmammadli commented:
@jdm I tried that but I don't know if I did it correctly or not

jdm commented:
Yeah, the result was messed up somehow. I cherry-picked the commit with your changes and force-pushed the result to your branch. If you make further contributions to Servo, I recommend following our [suggested workflow](https://github.com/servo/servo/wiki/Github-workflow). Thanks for fixing this!

@bors-servo r+

bors-servo commented:
:pushpin: Commit e6c85b9 has been approved by `jdm`

<!-- @bors-servo r=jdm e6c85b97f9420a8f1aa463f4cfc37d0db07d3a67 -->

bors-servo commented:
:hourglass: Testing commit e6c85b97f9420a8f1aa463f4cfc37d0db07d3a67 with merge eb09d8c6e2e0d84f5a43267a37614b27f3451311...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11073), [arm32](http://build.servo.org/builders/arm32/builds/11310), [arm64](http://build.servo.org/builders/arm64/builds/11343), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11381), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8262), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8277), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11437), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3831), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3859), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8105), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8172), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3932), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3879), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6772)
Approved by: jdm
Pushing eb09d8c6e2e0d84f5a43267a37614b27f3451311 to master...
